### PR TITLE
`LightningVJPs` works with new `LightningQubit`

### PR DESCRIPTION
**Pull request by mudit2812**
**Context:**
Lots of test cases were commented out once `LightningQubit` started using the new device API due to not having VJP support. This PR adds the needed changes for the `LightningQubit.compute_vjp` method to be fully integrated.

**Description of the Change:**
* Add logic to `LightningVJPs` to correctly handle VJP computation with devices using the new API.
* Uncomment interface tests using `LightningQubit` with `device_vjp=True`.
* Skip/xfail some unsupported tests. This mostly includes tests for state adjoint differentiation, which is not supported on `LightningQubit`.

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

**Comment by mudit2812** [comment]
This PR should not get merged until Lightning PR https://github.com/PennyLaneAI/pennylane-lightning/pull/644 is merged.

**Comment by mudit2812** [comment]
[sc-60397]

File: pennylane/workflow/jacobian_products.py [hunk 1]
```diff
@@ -741,10 +743,18 @@ def compute_vjp(self, tapes, dy):
...
-                numpy_tape.measurements, dyi, **self._processed_gradient_kwargs
-            )
-            out = vjp_f(numpy_tape)
+
+            if self._uses_new_device:
+                if self._execution_config is not None:
+                    out = self._device.compute_vjp(numpy_tape, dyi, self._execution_config)
+                else:
+                    out = self._device.compute_vjp(numpy_tape, dyi)
+            else:  # pragma: no cover
```
Comment (mudit2812) [inline pennylane/workflow/jacobian_products.py:752]: The `pragma: no cover` was adding to this line because the branch in question is used for old API devices (`lightning.kokkos` and `lightning.gpu`), which are not included in the testing suite.

File: pennylane/workflow/jacobian_products.py [hunk 2]
```diff
@@ -741,10 +743,18 @@ def compute_vjp(self, tapes, dy):
...
-            vjp_f = self._device.vjp(
-                numpy_tape.measurements, dyi, **self._processed_gradient_kwargs
-            )
-            out = vjp_f(numpy_tape)
+
+            if self._uses_new_device:
+                if self._execution_config is not None:
+                    out = self._device.compute_vjp(numpy_tape, dyi, self._execution_config)
+                else:
+                    out = self._device.compute_vjp(numpy_tape, dyi)
```
Comment (dwierichs) [inline pennylane/workflow/jacobian_products.py:751]: I understand that we need to separate out the execution kwarg if it is `None`, in order to not override the `DefaultExecutionConfig` in `_device.compute_vjp`. However, I'm wondering whether 
1. this is the most stable API for `compute_vjp`, or whether maybe the `DefaultExecutionConfig` should be passed down by default, confirming the default instead of creating a separate logical branch for it, and
2. there is a reason that we do not to do this for DQ? :thinking: Maybe I'm missing it and we actually are doing it, but I did not find it.

File: tests/devices/test_lightning_qubit.py [hunk 3]
```diff
@@ -89,42 +89,40 @@ def circuit():
     assert np.allclose(circ0(), circ1(), rtol=0.01)
 
 
-@pytest.mark.skip(reason="New lightning.qubit needs to be updated to work with dtypes")
```
Comment (dwierichs) [inline tests/devices/test_lightning_qubit.py:92]: Are the changes in this file related to the PR immediately, or more of a byproduct?

File: tests/interfaces/default_qubit_2_integration/test_autograd_qnode_default_qubit_2.py [hunk 4]
```diff
@@ -94,9 +94,9 @@ def circuit(a):
 
         assert circuit.qtape.interface is None
 
-        # without the interface, the QNode simply returns a scalar array
-        assert isinstance(res, np.ndarray)
-        assert res.shape == tuple()  # pylint: disable=comparison-with-callable
+        # without the interface, the QNode simply returns a scalar array or float
+        assert isinstance(res, (np.ndarray, float))
+        assert qml.math.shape(res) == tuple()  # pylint: disable=comparison-with-callable
```
Comment (dwierichs) [inline tests/interfaces/default_qubit_2_integration/test_autograd_qnode_default_qubit_2.py:99]: This is not optimal, I suppose? Shouldn't the QNode return value be independent of the device? :thinking: Sorry if this is an established behaviour, I may just not be used to it.

File: tests/interfaces/default_qubit_2_integration/test_autograd_qnode_default_qubit_2.py [hunk 5]
```diff
@@ -1319,6 +1319,9 @@ def cost_fn(x):
     def test_state(self, interface, dev, diff_method, grad_on_execution, device_vjp, tol):
         """Test that the state can be returned and differentiated"""
 
+        if "lightning" in getattr(dev, "name", "").lower():
+            pytest.xfail("Lightning does not support state adjoint diff.")
```
Comment (dwierichs) [inline tests/interfaces/default_qubit_2_integration/test_autograd_qnode_default_qubit_2.py:1323]: This is not necessary, I think, because below all tests with `diff_method!="backprop"` are `skip`ped anyways. If we want to switch to `xfail`, I would recommend that we do it for all devices.

File: tests/interfaces/default_qubit_2_integration/test_jax_jit_qnode_default_qubit_2.py [hunk 6]
```diff
@@ -1363,6 +1363,9 @@ def circuit(a, b):
     def test_state(self, dev, diff_method, grad_on_execution, device_vjp, interface, tol):
         """Test that the state can be returned and differentiated"""
 
+        if "lightning" in getattr(dev, "name", "").lower():
+            pytest.xfail("Lightning does not support state adjoint differentiation.")
```
Comment (dwierichs) [inline tests/interfaces/default_qubit_2_integration/test_jax_jit_qnode_default_qubit_2.py:1367]: As for Autograd.

File: tests/interfaces/default_qubit_2_integration/test_jax_qnode_default_qubit_2.py [hunk 7]
```diff
@@ -1242,6 +1242,9 @@ def circuit(a, b):
     def test_state(self, dev, diff_method, grad_on_execution, device_vjp, interface, tol):
         """Test that the state can be returned and differentiated"""
 
+        if "lightning" in getattr(dev, "name", "").lower():
+            pytest.xfail("Lightning does not support state adjoint differentiation.")
```
Comment (dwierichs) [inline tests/interfaces/default_qubit_2_integration/test_jax_qnode_default_qubit_2.py:1246]: As for Autograd

File: tests/interfaces/default_qubit_2_integration/test_torch_qnode_default_qubit_2.py [hunk 8]
```diff
@@ -1068,6 +1069,8 @@ def test_state(self, interface, dev, diff_method, grad_on_execution, device_vjp,
 
         if dev.name == "param_shift.qubit":
             pytest.skip("parameter shift does not support measuring the state.")
+        if "lightning" in getattr(dev, "name", "").lower():
+            pytest.xfail("Lightning devices do not support state with adjoint diff.")
```
Comment (dwierichs) [inline tests/interfaces/default_qubit_2_integration/test_torch_qnode_default_qubit_2.py:1073]: As for Autograd

File: tests/interfaces/test_jacobian_products.py [hunk 9]
```diff
@@ -67,10 +69,7 @@ def inner_execute_numpy(tapes):
     device_ps_jacs,
     device_native_jps,
     device_ps_native_jps,
-    pytest.param(
-        lightning_vjps,
-        marks=pytest.mark.skip(reason="Lighting JPC needs to be updated for new API"),
-    ),
+    lightning_vjps,
```
Comment (dwierichs) [inline tests/interfaces/test_jacobian_products.py:72]: :tada:

**Review by dwierichs — CHANGES_REQUESTED** [review]
Thanks @mudit2812, mostly looks good to me!
Just had some questions regarding the modified dtype test and the logic of the default execution config, as well as the strategy for introducing `xfail`s in some test.

Are we not supporting Tensorflow in general with Lightning device VJPs? If so, why is that (might be an ignorant question, sorry).

File: tests/interfaces/default_qubit_2_integration/test_autograd_qnode_default_qubit_2.py (hunk 5 above)
Comment (mudit2812) [reply tests/interfaces/default_qubit_2_integration/test_autograd_qnode_default_qubit_2.py:1323]: If `adjoint` diff is requested, the device executes and computes gradients in a single call rather than separate ones. So, this test fails with lightning before it reaches the `pytest.skip` line. The reason why only lightning tests are skipped is because `default.qubit` supports all state-based measurements with adjoint diff now.

File: tests/interfaces/default_qubit_2_integration/test_autograd_qnode_default_qubit_2.py (hunk 4 above)
Comment (mudit2812) [reply tests/interfaces/default_qubit_2_integration/test_autograd_qnode_default_qubit_2.py:99]: Good question. @albi3ro and I have discussed this briefly offline, and we decided it was okay to have `float` as a valid dtype for scalar returns with no interface.

Comment (dwierichs) [reply tests/interfaces/default_qubit_2_integration/test_autograd_qnode_default_qubit_2.py:99]: Fair enough! I hope this does not break any further processing, but as lightning got its own VJP method, I suppose it is fine :) Is this documented anywhere? :)

File: tests/interfaces/default_qubit_2_integration/test_autograd_qnode_default_qubit_2.py (hunk 5 above)
Comment (dwierichs) [reply tests/interfaces/default_qubit_2_integration/test_autograd_qnode_default_qubit_2.py:1323]: Oh, I see! So it is not about the differentiation being supported, but about the forward pass with a specific diff_method "in mind" being supported? Thanks! :)

File: pennylane/workflow/jacobian_products.py [hunk 10]
```diff
@@ -720,8 +720,10 @@ def __repr__(self):
         }
         return f"<LightningVJPs: {long_to_short_name[type(self._device).__name__]}, {self._gradient_kwargs}>"
 
-    def __init__(self, device, gradient_kwargs=None):
-        super().__init__(device, gradient_kwargs=gradient_kwargs)
+    def __init__(
+        self, device, gradient_kwargs=None, execution_config: "qml.devices.ExecutionConfig" = None
```
[the remark this answers is not in the document]
Comment (mudit2812) [reply pennylane/workflow/jacobian_products.py:724]: This is regular python syntax for objects not yet initialized. In this case, it is because of circular imports.

File: pennylane/workflow/jacobian_products.py (hunk 2 above)
Comment (mudit2812) [reply pennylane/workflow/jacobian_products.py:751]: The `DeviceJacobianProducts` class only works with devices using the new API, hence includes the `ExecutionConfig` as a required positional argument. However, `LightningVJPs` is supposed to work with both the old and new device API, since LGPU and LKokkos still use the old API, which does not work with `ExecutionConfig`s. I think it makes more sense to leave this as is for now, and then we can remove these branches, or maybe even this class altogether once LGPU and LKokkos are updated as well.

**Review by albi3ro — CHANGES_REQUESTED** [review]
`lightning.qubit` that implements the new device API should strictly use `DeviceJacobianProducts` and `DeviceJacobians`.  

`LightningVJPs` is strictly a patch to handle the vjp interface that was unique to the legacy lightning interface, and should now only apply to `lightning.kokkos` and `lightning.gpu`.

File: tests/devices/test_lightning_qubit.py (hunk 3 above)
Comment (mudit2812) [reply tests/devices/test_lightning_qubit.py:92]: It's a byproduct of changing `LightningQubit` to use the new API. Somewhere along processing measurements, the dtype gets converted to double precision. This needs to be investigated, but that's a task for another PR.

File: pennylane/workflow/jacobian_products.py (hunk 2 above)
Comment (mudit2812) [reply pennylane/workflow/jacobian_products.py:751]: Had some discussion with @albi3ro offline. Execution has been updated to use `DeviceJacobianProducts` with `lightning.qubit`.

File: tests/interfaces/test_jacobian_products.py [hunk 11]
```diff
@@ -225,6 +225,8 @@ def test_execute_jvp_basic(self, jpc, shots):
         """Test execute_and_compute_jvp for a simple single input single output."""
         if shots and not _accepts_finite_shots(jpc):
             pytest.skip("jpc does not work with finite shots.")
+        if isinstance(jpc, DeviceJacobianProducts) and "lightning" in jpc._device.name:
+            pytest.skip("Lightning devices don't have JVP method")
```
Comment (albi3ro) [inline tests/interfaces/test_jacobian_products.py:229]:
```suggestion
            pytest.xfail("Lightning devices don't have JVP method")
```
? Maybe.  Can't hurt and will let us know if things change.

**Review by dwierichs — APPROVED** [review]
Nice, @mudit2812 !

**Review by vincentmr — APPROVED** [review]
Thanks @mudit2812 , good to go on my side too.

File: pennylane/workflow/execution.py [hunk 12]
```diff
@@ -535,7 +535,7 @@ def cost_fn(params, x):
     if (
         device_vjp
         and isinstance(device, qml.Device)
-        and "lightning" not in getattr(device, "short_name", "")
+        and "lightning" not in getattr(device, "name", "").lower()
```
Comment (albi3ro) [inline pennylane/workflow/execution.py:538]:
```suggestion
        and "lightning" not in getattr(device, "short_name", "").lower()
```
This is now for qpu and kokkos which still implement the old device interface.

Comment (albi3ro) [reply pennylane/workflow/execution.py:538]: Though... I'm confused because we should still support VJPs for them through the `LightningVJPs` class?

Comment (mudit2812) [reply pennylane/workflow/execution.py:538]: I chose `name` because LGPU and LK have a `name` attribute along the lines of `"Lightning ... plugin for PennyLane"`, so it still works, and we won't have to change this again once they're ported to the new API.

Comment (mudit2812) [reply pennylane/workflow/execution.py:538]: > Though... I'm confused because we should still support VJPs for them through the LightningVJPs class?

This condition is for raising an error if we're requesting `device_vjp=True` with an old API device, and we don't want to include lightning devices in that list.

File: pennylane/workflow/jacobian_products.py [hunk 13]
```diff
@@ -678,7 +678,14 @@ def compute_vjp(self, tapes: Batch, dy: Tuple[Tuple[TensorLike]]) -> Tuple:
...
         dy = qml.math.unwrap(dy)
-        return self._device.compute_vjp(numpy_tapes, dy, self._execution_config)
+        vjps = self._device.compute_vjp(numpy_tapes, dy, self._execution_config)
+        res = []
+        for t, r in zip(tapes, vjps):
+            if len(t.trainable_params) == 1 and qml.math.shape(r) == ():
+                res.append((r,))
+            else:
+                res.append(r)
+        return res
```
Comment (albi3ro) [inline pennylane/workflow/jacobian_products.py:688]: Could we instead make this a constraint on `Device.compute_vjp`?  I think it it reasonable to add the constraint that the pytree structure of a VJP should match the structure of `tape.get_parameters()`.

Comment (mudit2812) [inline pennylane/workflow/jacobian_products.py:688]: I'm not particularly fond of this as it's doing the job of the device. We've specified the return type specification quite clearly, and devices should obey that. I've added this for now for the sake of getting the tests to pass, but long term I think it's better to update the Lightning device(s) to fix the return type.

Comment (mudit2812) [reply pennylane/workflow/jacobian_products.py:688]: Talked offline. We will leave this as is for now, update lightning to correctly handle vjp return types, and then remove this logic after that PR is merged.

Comment (mudit2812) [reply pennylane/workflow/jacobian_products.py:688]: Yep, I'll leave it like this for now, and we can update `LightningQubit` accordingly and revert this after that change is merged.

File: tests/interfaces/default_qubit_2_integration/test_autograd_qnode_default_qubit_2.py (hunk 4 above)
Comment (mudit2812) [reply tests/interfaces/default_qubit_2_integration/test_autograd_qnode_default_qubit_2.py:104]: Not currently. I will discuss with core team offline and update docs if we decide to.

Comment (mudit2812) [reply tests/interfaces/default_qubit_2_integration/test_autograd_qnode_default_qubit_2.py:104]: @albi3ro  has a doc PR that will address this.

**Review by albi3ro — APPROVED** [review]
Thanks for the changes 👍
